### PR TITLE
Migratable SR-IOV logical device test

### PR DESCRIPTION
**Pull request by abdhaleegit**
Code changes adds support for HNV sriov test, functions inlcudes

1. Create Migratable SRIOV logical device with vETH as backup device
2. Create Migratable SRIOV logical device with vNIC as backup device
3. Configure the HNV hcn bond device and perform ping check
4. Delete the HNV hcn bond device
5. Delete the Migratable SRIOV logical device

Made sure the existing SRIOV logic works and also above test can be
run explicitly, have added separate yamls for veth and vnic tests

Signed-off-by: Abdul Haleem <abdhalee@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by smruti77** [comment]
@abdhaleegit Please upload job.log file.

File: io/net/sriov_device_test.py [hunk 1]
```diff
@@ -169,6 +251,32 @@ def get_logical_port_id(self, mac):
         logical_port_id = output.stdout_text.split(',')[6].split('=')[-1]
         return logical_port_id
 
+    def get_hnv_bond(self):
+        """
+        Get the newly created hnv bond interface name
+        """
+        output = process.system_output('cat /sys/class/net/bonding_masters',
+                                       ignore_status=True, shell=True,
+                                       sudo=True).decode("utf-8").split()
```
Comment (narasimhan-v) [inline io/net/sriov_device_test.py:260]: `genio.read_file()`

File: io/net/sriov_device_test.py [hunk 2]
```diff
@@ -169,6 +251,32 @@ def get_logical_port_id(self, mac):
...
+        for bond in output:
+            if self.mac_id in netifaces.ifaddresses(bond)[17][0]['addr']:
+                return bond
+        return ''
+
+    def bringup_hnv_bond(self, bond_name):
+        """
+        Configure hnv bond interface with nmcli command
```
Comment (narasimhan-v) [inline io/net/sriov_device_test.py:268]: Should we check nmcli package installation too ?

**Review by narasimhan-v — CHANGES_REQUESTED** [review]

File: io/net/sriov_device_test.py (hunk 2 above)
Comment (abdhaleegit) [reply io/net/sriov_device_test.py:268]: yes, I will add a check for NetworkManager pacakge

File: io/net/sriov_device_test.py (hunk 1 above)
Comment (abdhaleegit) [reply io/net/sriov_device_test.py:260]: Yeap, I replaced with genio way of reading file

File: io/net/sriov_device_test.py [hunk 3]
```diff
@@ -70,12 +71,48 @@ def setUp(self):
...
+        self.backup_vnic_backing_device = 'sriov/%s/%s/%s/%s/%s/%s/%s' % (
+            self.vios_name, self.vios_id, self.vnic_adapter_id, self.vnic_port_id, self.capacity, self.priority, self.max_capacity)
         self.local = LocalHost()
 
+    @staticmethod
+    def netmask_to_cidr(netmask):
+        return(sum([bin(int(bits)).count("1") for bits in netmask.split(".")]))
+
+    @staticmethod
+    def get_adapter_id(slot):
```
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply io/net/sriov_device_test.py:107]: yes canceling test if adapter not found

File: io/net/sriov_device_test.py [hunk 4]
```diff
@@ -70,12 +71,48 @@ def setUp(self):
...
+              ' -r lpar --filter lpar_names=' + self.vios_name + \
+              ' -F lpar_id'
+        self.vios_id = self.session.cmd(cmd).stdout_text.split()[0]
+        self.vnic_sriov_adapter = self.params.get(
+            'vnic_sriov_adapter', '*', default=None)
+        self.vnic_port_id = self.params.get('vnic_port_id', '*', default=None)
+        self.vnic_adapter_id = self.get_adapter_id(self.vnic_sriov_adapter)
+        self.priority = self.params.get('failover-priority', '*', default='50')
+        self.backup_vnic_backing_device = 'sriov/%s/%s/%s/%s/%s/%s/%s' % (
+            self.vios_name, self.vios_id, self.vnic_adapter_id, self.vnic_port_id, self.capacity, self.priority, self.max_capacity)
```
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply io/net/sriov_device_test.py:99]: yes, made it multiline with in 80 charachters

**Comment by abdhaleegit** [comment]
```
avocado run sriov_device_test.py --mux-yaml sriov_device_test.py.data/migratable_sriov_veth.yaml
JOB ID     : f2887b4ac0938be401762d1857dd1028097107b3
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2020-11-26T07.39-f2887b4/job.log
 (1/4) sriov_device_test.py:NetworkSriovDevice.test_add_logical_device;run-5bd7: CANCEL: Test unsupported (1.63 s)
 (2/4) sriov_device_test.py:NetworkSriovDevice.test_add_migratable_sriov;run-5bd7: FAIL: Test fail due to mac address mismatch (12.41 s)
 (3/4) sriov_device_test.py:NetworkSriovDevice.test_remove_migratable_sriov;run-5bd7: FAIL: Test fail due to mac address mismatch (1.72 s)
 (4/4) sriov_device_test.py:NetworkSriovDevice.test_remove_logical_device;run-5bd7: CANCEL: Test unsupported (1.67 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2020-11-26T07.39-f2887b4/results.html
JOB TIME   : 18.33 s
```
test fails due to real linux bug open

File: io/net/sriov_device_test.py [hunk 5]
```diff
@@ -70,12 +72,49 @@ def setUp(self):
...
+
+    @staticmethod
+    def get_adapter_id(slot):
+        cmd = "lshwres -m %s -r sriov --rsubtype adapter -F phys_loc:adapter_id" \
+              % (self.server)
+        output = self.session.cmd(cmd)
+        for line in output.stdout_text.splitlines():
+            if slot in line:
+                return line.split(':')[-1]
+        self.cancel("adapter not found at slot %s", slot)
```
Comment (narasimhan-v) [inline io/net/sriov_device_test.py:116]: Is it fine to use self in a staticmethod ?

**Review by narasimhan-v — CHANGES_REQUESTED** [review]

**Comment by narasimhan-v** [comment]
@abdhaleegit I guess Travis is also complaining about the self use in static method.

File: io/net/sriov_device_test.py (hunk 5 above)
Comment (abdhaleegit) [reply io/net/sriov_device_test.py:116]: looks like I havent pushed latest code

**Comment by abdhaleegit** [comment]
Migratable Sriov logs with VETH as Backing device 

```
JOB ID     : d96d0c97adad694b8d0cbc43af01adaaa5678130
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2020-11-26T12.08-d96d0c9/job.log
 (1/4) sriov_device_test.py:NetworkSriovDevice.test_add_logical_device;run-5bd7: CANCEL: Test unsupported (1.76 s)
 (2/4) sriov_device_test.py:NetworkSriovDevice.test_add_migratable_sriov;run-5bd7: PASS (16.94 s)
 (3/4) sriov_device_test.py:NetworkSriovDevice.test_remove_migratable_sriov;run-5bd7: PASS (15.04 s)
 (4/4) sriov_device_test.py:NetworkSriovDevice.test_remove_logical_device;run-5bd7: CANCEL: Test unsupported (1.60 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2020-11-26T12.08-d96d0c9/results.html
JOB TIME   : 36.12 s
```
[msriovvethlogs.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/5604354/msriovvethlogs.txt)

**Comment by abdhaleegit** [comment]
Migratable SRIOV logs with vNIC as backup device
```
[root@ltcfleet2-lp6 net]# avocado run sriov_device_test.py --mux-yaml sriov_device_test.py.data/migratable_sriov_vnic.yaml
JOB ID     : b0e15c48d9bab4ddb7ce443575aaf9cdcc8064a4
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2020-11-26T12.36-b0e15c4/job.log
 (1/4) sriov_device_test.py:NetworkSriovDevice.test_add_logical_device;run-dd6f: CANCEL: Test unsupported (1.74 s)
 (2/4) sriov_device_test.py:NetworkSriovDevice.test_add_migratable_sriov;run-dd6f: PASS (29.93 s)
 (3/4) sriov_device_test.py:NetworkSriovDevice.test_remove_migratable_sriov;run-dd6f: PASS (23.22 s)
 (4/4) sriov_device_test.py:NetworkSriovDevice.test_remove_logical_device;run-dd6f: CANCEL: Test unsupported (1.70 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2020-11-26T12.36-b0e15c4/results.html
JOB TIME   : 57.38 s
```
[msriovvniclogs.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/5604425/msriovvniclogs.txt)

**Comment by abdhaleegit** [comment]
Just SRIOV logical device tests logs

```
[root@ltcfleet2-lp6 net]# avocado run sriov_device_test.py --mux-yaml sriov_device_test.py.data/sriov_device_test.yaml
JOB ID     : 4b275d194ef54166b08d995e8596e35cbe3f9c71
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2020-11-26T12.52-4b275d1/job.log
 (1/4) sriov_device_test.py:NetworkSriovDevice.test_add_logical_device;run-81b3: PASS (15.42 s)
 (2/4) sriov_device_test.py:NetworkSriovDevice.test_add_migratable_sriov;run-81b3: CANCEL: Test unsupported (1.69 s)
 (3/4) sriov_device_test.py:NetworkSriovDevice.test_remove_migratable_sriov;run-81b3: CANCEL: Test unsupported (1.66 s)
 (4/4) sriov_device_test.py:NetworkSriovDevice.test_remove_logical_device;run-81b3: PASS (13.39 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 2
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2020-11-26T12.52-4b275d1/results.html
JOB TIME   : 32.96 s
```
[justsriovlogs.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/5604494/justsriovlogs.txt)

**Comment by narasimhan-v** [comment]
@vaishnavibhat Review comments please.

**Review by narasimhan-v — APPROVED** [review]

**Comment by abdhaleegit** [comment]
@narasimhan-v @vaishnavibhat any thing pending to merge ??

**Pull request merged**